File: java/src/main/java/com/example/design_pattern/singleton/SingletonHolder.java
package com.example.design_pattern.singleton;

public class SingletonHolder {
	private static class Database {
		private Database(){}

		private static class DatabaseHolder{
			public static final Database db = new Database();
		}

		public static Database getInstance(){
			return DatabaseHolder.db;
		}
	}

	public static void main(String... args){
		Database db1 = Database.getInstance();
		Database db2 = Database.getInstance();

		assert db1 == db2;
	}
}

File: typescript/design_pattern/singleton/singletonHolder.ts
class Database {
	private static DatabaseHolder = {
		db : new Database()
	}

	private constructor() {}

	static getInstance() {
		return Database.DatabaseHolder.db
	}
}

function main() {
	const db1 = Database.getInstance();
	const db2 = Database.getInstance();

	console.log(db1 === db2)
}

main()